Linking fails for Windows on nightly rust (and future stable versions)

As of https://github.com/rust-lang/rust/pull/138233 `mimalloc` no longer compiles successfully on windows systems.

Original issue (closed as duplicate) on rust-lang repository here: https://github.com/rust-lang/rust/issues/140041

As far as I can tell this is a change they intend to keep and it is expected for crate authors to adapt.


#### Backtrace

```
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Users\\Admin\\Programs\\VisualStudio\\2022\\Community\\VC\\Tools\\MSVC\\14.43.34808\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "C:\\Users\\Admin\\AppData\\Local\\Temp\\rustc76KEBa\\symbols.o" "<2 object files omitted>" "C:\\Users\\Admin\\Projects\\test_mimalloc\\target\\release\\deps/{libmimalloc-ad67d45c7a502e86.rlib,liblibmimalloc_sys-b9c6bcbae022bf4a.rlib,liblibc-fa9e52f7b8cff268.rlib}.rlib" "<sysroot>\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib/{libstd-*,libpanic_unwind-*,libwindows_targets-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libunwind-*,libcfg_if-*,liballoc-*,librustc_std_workspace_core-*,libcore-*,libcompiler_builtins-*}.rlib"
"legacy_stdio_definitions.lib" "kernel32.lib" "kernel32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib"
"dbghelp.lib" "/defaultlib:msvcrt" "/NXCOMPAT" "/LIBPATH:C:\\Users\\Admin\\Programs\\VisualStudio\\2022\\Community\\VC\\Tools\\MSVC\\14.43.34808\\atlmfc\\lib\\x64" "/LIBPATH:C:\\Users\\Admin\\Projects\\test_mimalloc\\target\\release\\build\\libmimalloc-sys-3b898ff4937b9f62\\out" "/OUT:C:\\Users\\Admin\\Projects\\test_mimalloc\\target\\release\\deps\\test_mimalloc.exe" "/OPT:REF,ICF" "/DEBUG" "/PDBALTPATH:%_PDB%" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\libstd.natvis"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: liblibmimalloc_sys-b9c6bcbae022bf4a.rlib(5a07bf3761bb5df8-static.o) : error LNK2019: unresolved external symbol __imp_OpenProcessToken referenced in function win_enable_large_os_pages␍
          liblibmimalloc_sys-b9c6bcbae022bf4a.rlib(5a07bf3761bb5df8-static.o) : error LNK2019: unresolved external symbol __imp_AdjustTokenPrivileges referenced in function win_enable_large_os_pages␍
          liblibmimalloc_sys-b9c6bcbae022bf4a.rlib(5a07bf3761bb5df8-static.o) : error LNK2019: unresolved external symbol __imp_LookupPrivilegeValueA referenced in function win_enable_large_os_pages␍
          C:\Users\Admin\Projects\test_mimalloc\target\release\deps\test_mimalloc.exe : fatal error
LNK1120: 3 unresolved externals
```

--- a/libmimalloc-sys/build.rs
+++ b/libmimalloc-sys/build.rs
@@ -63,4 +63,14 @@ fn main() {
         let atomic_name = env::var("DEP_ATOMIC").unwrap_or("atomic".to_owned());
         println!("cargo:rustc-link-lib={}", atomic_name);
     }
+
+    // Link with libs needed on Windows
+    if target_os == "windows" {
+        // https://github.com/microsoft/mimalloc/blob/af21001f7a65eafb8fb16460b018ebf9d75e2ad8/CMakeLists.txt#L487
+        let libs = ["psapi", "shell32", "user32", "advapi32", "bcrypt"];
+
+        for lib in libs {
+            println!("cargo:rustc-link-lib={}", lib);
+        }
+    }
 }

